File: leetcode/684. Redundant Connection/Solution.py
class Solution:
    def __init__(self):
        self.ranks = []
        self.store = []

    def findRedundantConnection(self, edges):
        """
        :type edges: List[List[int]]
        :rtype: List[int]
        --------------------
        39 / 39 test cases passed.
        Status: Accepted
        Runtime: 44 ms
        """
        self.store = [i for i in range(len(edges) * 2)]
        self.ranks = [1] * (len(edges) * 2)
        for e in edges:
            if not self.__union(e[0], e[1]):
                return e
        return [0, 0]

    def __find(self, p):
        while p != self.store[p]:
            self.store[p] = self.store[self.store[p]]
            p = self.store[p]
        return p

    def __union(self, p, q):
        rootP = self.__find(p)
        rootQ = self.__find(q)
        if rootP == rootQ:
            return False
        if self.ranks[rootP] < self.ranks[rootQ]:
            self.store[rootP] = rootQ
        else:
            self.store[rootQ] = rootP
            if self.ranks[rootQ] == self.ranks[rootP]:
                self.ranks[rootP] += 1
        return True


if __name__ == '__main__':
    # print(Solution().findRedundantConnection([[1,2],[1,3],[2,3]]))
    print(Solution().findRedundantConnection([[16,25],[7,9],[3,24],[10,20],[15,24],[2,8],[19,21],[2,15],[13,20],[5,21],[7,11],[6,23],[7,16],[1,8],[17,20],[4,19],[11,22],[5,11],[1,16],[14,20],[1,4],[22,23],[12,20],[15,18],[12,16]]))

File: LeetCode/684. Redundant Connection/Solution.java
/**
 * Created by Inno Fang on 2018/3/28.
 */

/**
 * 39 / 39 test cases passed.
 * Status: Accepted
 * Runtime: 5 ms
 */
class Solution {
    private int[] store;
    private int[] ranks;

    private int find(int p) {
        while (p != store[p]) {
            p = store[p] = store[store[p]];
        }
        return p;
    }

    private boolean union(int p, int q) {
        int rootP = find(p);
        int rootQ = find(q);
        if (rootP == rootQ) return false;
        if (ranks[rootP] < ranks[rootQ]) store[rootP] = rootQ;
        else {
            store[rootQ] = rootP;
            if (ranks[rootQ] == ranks[rootP]) ranks[rootP]++;
        }
        return true;
    }

    public int[] findRedundantConnection(int[][] edges) {
        int num = edges.length * 2;
        store = new int[num];
        ranks = new int[num];
        for (int i = 0; i < num; i++) {
            store[i] = i;
            ranks[i] = 1;
        }
        for (int[] edge : edges) {
            if (!union(edge[0], edge[1])) {
                return edge;
            }
        }
        return new int[2];
    }
}